tools/codegen/model_generator/model.go

```diff
@@ -0,0 +1,257 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model_generator
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	tmplExtns "istio.io/mixer/tools/codegen/template_extension"
+	"github.com/golang/protobuf/proto"
+)
+
+/*
+Data model to be used to generate code for istio/mixer
+*/
+type Model struct {
+	// top level fields
+	Name        string
+	Check       bool
+	PackageName string
+	VarietyName string
+
+	// imports
+	Imports []string
+
+	// Constructor fields
+	ConstructorFields []FieldInfo
+}
+
+type FieldInfo struct {
+	Name string
+	Type TypeInfo
+}
+
+type TypeInfo struct {
+	Name string
+}
+
+const FullNameOfExprMessage = "istio_mixer_v1_config_template.Expr"
+const FullNameOfExprMessageWithPtr = "*" + FullNameOfExprMessage
+
+// Creates a Model object.
+func CreateModel(parser *FileDescriptorSetParser) (Model, error) {
+	model := &Model{}
+
+	templateProto, err := getTmplFileDesc(parser.allFiles)
+	if err != nil {
+		return *model, err
+	}
+	if !templateProto.proto3 {
+		return *model, fmt.Errorf("Only proto3 template files are allowed")
+	}
+	// set the current generated code package to the package of the
+	// templateProto. This will make sure references within the
+	// generated file into the template's pb.go file are fully qualified.
+	parser.packageName = goPackageName(templateProto.GetPackage())
+
+	err = model.addTopLevelFields(templateProto)
+	if err != nil {
+		return *model, err
+	}
+
+	// ensure Constructor is present
+	cnstrDesc, err := getRequiredMsg(templateProto, "Constructor")
+	if err != nil {
+		return *model, err
+	}
+
+	// ensure Constructor is present
+	_, err = getRequiredMsg(templateProto, "Type")
+	if err != nil {
+		return *model, err
+	}
+
+	err = model.addInstanceFieldFromConstructor(parser, templateProto)
+	if err != nil {
+		return *model, err
+	}
+
+	// imports only referenced by the Constructor message.
+	model.addImports(parser, templateProto, cnstrDesc)
+	if err != nil {
+		return *model, err
+	}
+
+	return *model, err
+}
+
+// Find the file that has the options TemplateVariety and TemplateName. There should only be one such file.
+func getTmplFileDesc(fds []*FileDescriptor) (*FileDescriptor, error) {
+	var templateDescriptorProto *FileDescriptor = nil
+
+	for _, fd := range fds {
+		if fd.GetOptions() == nil {
+			continue
+		}
+		if !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) && !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
+			continue
+		} else if proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) && proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
+			if templateDescriptorProto == nil {
```

Comment:
This function can be simplified somewhat:

		if !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) && !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
			continue
		}

                if !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateName) || !proto.HasExtension(fd.GetOptions(), tmplExtns.E_TemplateVariety) {
			return nil, fmt.Errorf("Proto files %s has only one of the "+
				"following two options %s and %s. Both options are required.",
				fd.GetName(), tmplExtns.E_TemplateVariety.Name, tmplExtns.E_TemplateName.Name)
                }

		if templateDescriptorProto != nil {
			return nil, fmt.Errorf("Proto files %s and %s, both have"+
				" the options %s and %s. Only one proto file is allowed with those options",
				fd.GetName(), templateDescriptorProto.Name,
				tmplExtns.E_TemplateVariety.Name, tmplExtns.E_TemplateName.Name)
                }
		templateDescriptorProto = fd